### PR TITLE
Formalize finite implications from 3342

### PR DESCRIPTION
**Pull request by vlad902**
Two small notes to the reviewers:
- This is my first time using `\lean` and `\leanok` in the blueprint, it seems straightforward but please check that I'm using it correctly.
- `Finite.fn_eventually_periodic` (and maybe `fn_periodic`) will probably be moved to somewhere common for use in #809, I included it with the proofs for 3342 for now since I'm not done with that proof yet and not sure what it will look like in the end.

### CODE REVIEW

File: equational_theories/ManuallyProved/Equation3342.lean [hunk 1]
```diff
@@ -58,4 +63,106 @@ theorem Equation3342_facts : ∃ (G : Type) (_ : Magma G), Facts G [3342] [3456,
...
+    refine ⟨p, hpgt, this⟩
+  . let n := s * p
+    have : f^[n] = f^[2*n] := by
+      unfold n
+      obtain ⟨ppred, hppred⟩ := Nat.exists_eq_succ_of_ne_zero (by linarith)
+      rw [hppred, Nat.succ_eq_add_one, Nat.mul_succ, Nat.add_comm]
+      have : 2 * (s + s * ppred) = s + s * ppred + s * p := by simp_arith only [hppred,  Nat.mul_succ]
+      rw [this, ← hmod]
+    have ngt : n > 0 := by apply Nat.mul_pos h hpgt
+    refine ⟨n, ngt, this⟩
```
Comment (madvorak) [inline equational_theories/ManuallyProved/Equation3342.lean:100]:
```suggestion
    exact ⟨n, ngt, this⟩
```

File: equational_theories/ManuallyProved/Equation3342.lean [hunk 2]
```diff
@@ -58,4 +63,106 @@ theorem Equation3342_facts : ∃ (G : Type) (_ : Magma G), Facts G [3342] [3456,
...
+
+lemma main_result (G : Type*) [Magma G] [Finite G] (h : Equation3342 G) : Equation3522 G ∧ Equation4118 G := by
```
Comment (madvorak) [inline equational_theories/ManuallyProved/Equation3342.lean:102]: This is a weird decomposition into lemmas, but I can live with it.

File: equational_theories/ManuallyProved/Equation3342.lean [hunk 3]
```diff
@@ -58,4 +63,106 @@ theorem Equation3342_facts : ∃ (G : Type) (_ : Magma G), Facts G [3342] [3456,
     use (0, 0), (0, 0)
     decide
 
+-- https://leanprover.zulipchat.com/#narrow/channel/458659-Equational/topic/Austin.20pairs/near/482525422
+
+lemma Finite.fn_periodic (G : Type*) [Finite G] (f : G -> G) : ∃ p : ℕ, p > 0 ∧ f^[p] = f^[2*p] := by
+  have Finite.fn_eventually_periodic : ∃ s p : ℕ, p > 0 ∧ f^[s] = f^[s+p] := by
+    obtain ⟨p1, p2, lt, heq⟩ : ∃ p1 p2 : ℕ, p1 < p2 ∧ f^[p1] = f^[p2] := by
```
Comment (madvorak) [inline equational_theories/ManuallyProved/Equation3342.lean:70]: Write lower indices please.

\1
\2

**Comment by madvorak** [comment]
Congratulations!

File: equational_theories/ManuallyProved/Equation3342.lean (hunk 2 above)
Comment (vlad902) [reply equational_theories/ManuallyProved/Equation3342.lean:102]: If you mean that there are two lemmas (fn_periodic and main_result), the first lemma will probably be reused in a future proof so it was easier to keep it separate. If you mean the fact that this lemma proves both equations, I agree that it's weird, but trying to construct it so the equation-specific proofs were in the separate implications led to requiring some of the internal hypotheses to be split out and that led to thers needing to be separate and it seemed worse than this approach in the end, but I'm a Lean-beginner and very open to feedback on how to structure it better!

**Comment by vlad902** [comment]
Thank you for the feedback!

File: equational_theories/ManuallyProved/Equation3342.lean (hunk 2 above)
Comment (madvorak) [reply equational_theories/ManuallyProved/Equation3342.lean:102]: You know better has to structure your proof than I do.

I just know the common pattern — when two proofs share a lot of code, it is better to extract the shared part into a lemma than to merge the two theorems in order to have the shared parts only once. You can also use ˙private lemma˙ if you want.

File: equational_theories/ManuallyProved/Equation3342.lean [hunk 4]
```diff
@@ -58,4 +63,107 @@ theorem Equation3342_facts : ∃ (G : Type) (_ : Magma G), Facts G [3342] [3456,
...
+    have : f^[n] = f^[2*n] := by
+      unfold n
+      obtain ⟨ppred, hppred⟩ := Nat.exists_eq_succ_of_ne_zero (by linarith)
+      rw [hppred, Nat.succ_eq_add_one, Nat.mul_succ, Nat.add_comm]
+      have : 2 * (s + s * ppred) = s + s * ppred + s * p := by simp_arith only [hppred,  Nat.mul_succ]
+      rw [this, ← hmod]
+    have ngt : n > 0 := by apply Nat.mul_pos h hpgt
+    exact ⟨n, ngt, this⟩
+
+private lemma main_result (G : Type*) [Magma G] [Finite G] (h : Equation3342 G) :
```
Comment (madvorak) [inline equational_theories/ManuallyProved/Equation3342.lean:102]: What I meant is that, if you wanted, you could make `private lemma` out of those `have` inside this proof and then refer to them in both theorems.

I understand if you don't feel like refactoring it. In this case, I'd just change `private lemma main_result` to `private theorem main_result` and merge it soon.

Comment (vlad902) [reply equational_theories/ManuallyProved/Equation3342.lean]: Yeah, I just gave it a go and gave up again. You need to pass around the period and related hypotheses and it made it less clear in the end, even if it was more modular. I've just made the lemma->theorem change you recommend.

**Comment by madvorak** [comment]
LGTM, somebody just check the `leanok` which I don't know how to use either.

File: blueprint/src/chapter/infinite_models.tex [hunk 5]
```diff
@@ -128,18 +128,20 @@ \chapter{Infinite models}\label{infinite-model-chapter}
 This can be used to obtain a few positive finite magma implications, for instance by setting $f,g$ to be left and right multiplication operators.  Another useful lemma is
 
 \begin{lemma}[Eventual period]\label{period}  Let $X$ be finite and $f: X \to X$.  Then there exists $n \geq 1$ such that $f^{2n} = f^n$.
+  \lean{Eq3342.Finite.fn_periodic}\leanok
```
Comment (pitmonticone) [inline blueprint/src/chapter/infinite_models.tex:131]:
```suggestion
\begin{lemma}[Eventual period]\label{period} \lean{Eq3342.Finite.fn_periodic} \leanok
    Let $X$ be finite and $f: X \to X$.  Then there exists $n \geq 1$ such that $f^{2n} = f^n$.
```

**Comment by vlad902** [comment]
I've made the mentioned updates but I've been stuck making the build work. My limited understanding is that the recent mathlib/lean bump is responsible for it, but I don't fully understand it. Previously, e.g. in the following example:
```
example {G : Type*} [Finite G] (f : G → G) : False := by
  obtain ⟨p₁, p₂, ne, heq⟩ := Finite.exists_ne_map_eq_of_infinite (Nat.iterate f ·)
```
[Finite G -> G] was automatically synthesized, and now it's not. Perhaps this is due to lean4#6024, I'm trying to debug.

**Comment by vlad902** [comment]
The cause was much more prosaic: I believe mathlib 9a9cb1894400637ebac1c1c8ac2702c23412ee5e simply split some imports I was previously transitively depending on, and I no longer imported the instance I needed.

**Pull request merged**